syntax = "proto3";

package plagarism_detection;

service PlagarismDetectionService {
    rpc Ping(PingRequest) returns (PingResponse) {}
    rpc Login(LoginRequest) returns (LoginResponse) {}
    rpc Register(RegisterRequest) returns (RegisterResponse) {}
    rpc UploadFile(UploadFileRequest) returns (UploadFileResponse) {}
    rpc CheckTwoFiles(CheckTwoFilesRequest) returns (CheckTwoFilesResponse) {}
}

message PingRequest {
}

message PingResponse {
    int32 status = 1;
}

message LoginRequest {
    string username = 1;
    string password = 2;
}

message LoginResponse {
    LoginStatus status = 1;
    string token = 2;
}

message RegisterRequest {
    string username = 1;
    string password = 2;
}

message RegisterResponse {
    RegisterStatus status = 1;
    optional string token = 2;
}

message UploadFileRequest {
    string token = 1;
    string file = 2;
}

message UploadFileResponse {
    string token = 1;
    string file_id = 2;
}

message CheckTwoFilesRequest {
    string token = 1;
    string file_id_1 = 2;
    string file_id_2 = 3;
}

message CheckTwoFilesResponse {
    repeated int32 line_numbers = 1;
}

enum LoginStatus {
    LOGIN_SUCCESS = 0;
    LOGIN_USERNAME_EMPTY = 1;
    LOGIN_PASSWORD_EMPTY = 2;
    LOGIN_INVALID_CREDENTIALS = 3;
    LOGIN_UNKNOWN_ERROR = 4;
}

enum RegisterStatus {
    REGISTER_SUCCESS = 0;
    REGISTER_USERNAME_EMPTY = 1;
    REGISTER_USERNAME_INVALID = 2;
    REGISTER_PASSWORD_EMPTY = 3;
    REGISTER_PASSWORD_INVALID = 4;
    REGISTER_PASSWORDS_MISMATCH = 5;
    REGISTER_USERNAME_TAKEN = 6;
    REGISTER_UNKNOWN_ERROR = 7;
}

